Convert Vernacular section of gallina chapter to use prodn (#11423)
coq/coq: doc/sphinx/language/gallina-specification-language.rst

Comments
  Comments are enclosed between ``(*`` and ``*)``.  They can be nested.

>   They can contain any character. However, embedded :n:`@string` literals must be

Author: I was following up on https://github.com/coq/coq/pull/11423#discussion_r370023798
though I was only looking for :token:\`XXX\` and not subscripts.

There are still a lot of these :token:s.  I think they look better when converted to :n:\`@xxx\` - you get the pretty green type for the nonterminal whereas tokens appear in black.

>   They can contain any character. However, embedded :n:`@string` literals must be

Reviewer: I thought we had changed the output of `:token:` so that it would be identical to `:n:` with `@`.

>   They can contain any character. However, embedded :n:`@string` literals must be

Author: You're right, they do look the same, both are green.

   | ( {+ @name } : @term )
   | ( @name {? : @term } := @term )
   | %{ {+ @name } {? : @term } %}
   | [ @name ]
   | [ @name {+ @name } : @term ]
   | [ @name : @term ]

>    | [ @name {+ @name } ]

Author: This was an external change to the .mlg file.  See `fullGrammar`.

I'll condense this to 2 lines by making `: @term` optional.

Too many kinds of brackets here for my taste: ( [ { \`( \`[ \`{.

each terminated with a period
and optionally decorated with :ref:`gallina-attributes`.  :n:`@ltac_expr` syntax supports both simple
and compound tactics.  For example: ``split.`` is a simple tactic while ``split; auto.`` combines two
simple tactics.

Tactics specify how to transform the current proof state as a step in creating a proof.  They
are syntactically valid only when |Coq| is in proof mode, such as after a :cmd:`Theorem` command
and before any subsequent proof-terminating command such as :cmd:`Qed`.  See :ref:`proofhandling` for more
on proof mode.

By convention, command names begin with uppercase letters, while
tactic names begin with lowercase letters.  Commands appear in the
HTML documentation in blue boxes after the label "Command".  In the pdf, they appear
after the boldface label "Command:".  Commands are listed in the :ref:`command_index`.

Similarly, tactics appear after the label "Tactic".  Tactics are listed in the :ref:`tactic_index`.

.. todo:: What about removing the proof part of this grammar from this chapter
          and putting it somewhere where top-level tactics can be described as well?

>           See also #7583.

Reviewer: I don't think this needs to be kept in the final version.

```suggestion
```

   | @ @qualid {? @univ_annot }

The expression “``fix`` :n:`@ident`:math:`_1` :n:`@binder`:math:`_1` ``:``
:n:`@type`:math:`_1` ``:=`` :n:`@term`:math:`_1` ``with … with``
:n:`@ident`:math:`_n` :n:`@binder`:math:`_n` : :n:`@type`:math:`_n`
``:=`` :n:`@term`:math:`_n` ``for`` :n:`@ident`:math:`_i`” denotes the
:math:`i`-th component of a block of functions defined by mutual structural
recursion. It is the local counterpart of the :cmd:`Fixpoint` command. When

> :math:`n=1`, the “``for`` :n:`@ident`:math:`_i`” clause is omitted.

Reviewer: That's the kind of change I was suggesting in [this comment](https://github.com/coq/coq/pull/11423#discussion_r370023798).

   These commands bind one or more :n:`@ident`\(s) to specified :n:`@type`\(s) as their specifications in
   the global context. The fact asserted by the :n:`@type` is accepted as a
   postulate.

   :cmd:`Axiom`, :cmd:`Conjecture`, :cmd:`Parameter` and their plural forms
   are equivalent.  They can take the :attr:`local` attribute (see :ref:`gallina-attributes`),
   which makes the defined :n:`@ident`\s accessible by :cmd:`Import` and its variants
   only through their fully qualified names.

   Similarly, :cmd:`Hypothesis`, :cmd:`Variable` and their plural forms are equivalent.  Outside
   of a section, these are equivalent to :n:`Local Parameter`.  Inside a section, the
   :n:`@ident`\s defined are only accessible within the section.  When the current section
   is closed, the :n:`@ident`\(s) become undefined and every object depending on them will be explicitly
   parameterized (i.e., the variables are *discharged*).  See Section :ref:`section-mechanism`.

.. example:: Simple assumptions

    .. coqtop:: reset in

       Parameter X Y : Set.
       Parameter (R : X -> Y -> Prop) (S : Y -> X -> Prop).
       Axiom R_S_inv : forall x y, R x y <-> S y x.

> .. comment

Author: Oops.  I wanted to get rid of the extra space at the bottom of the example but this was not the way to do it.

![image](https://user-images.githubusercontent.com/1253341/74760178-dd1f2300-522e-11ea-9eb5-9f9a79764cf6.png)

> .. comment

Reviewer: I have no idea how to fix this, but probably not something to worry about for the moment.

   | cofix @cofix_body {? {+ with @cofix_body } for @ident }
   cofix_body ::= @ident {* @binder } {? : @term } := @term

The expression “``cofix`` :n:`@ident`:math:`_1` :n:`@binder`:math:`_1` ``:``
:n:`@type`:math:`_1` ``with … with`` :n:`@ident`:math:`_n` :n:`@binder`:math:`_n`
: :n:`@type`:math:`_n` ``for`` :n:`@ident`:math:`_i`” denotes the
:math:`i`-th component of a block of terms defined by a mutual guarded
co-recursion. It is the local counterpart of the :cmd:`CoFixpoint` command. When
:math:`n=1`, the “``for`` :n:`@ident`:math:`_i`” clause is omitted.

.. _vernacular:

The Vernacular
==============

.. insertprodn vernacular vernacular

.. prodn::
   vernacular ::= {* {? @all_attrs } {| @command | @ltac_expr } . }

The top-level input to |Coq| is a series of commands and tactics, each terminated with a period

> and optionally decorated with :ref:`gallina-attributes`.  :n:`@ltac_expr` syntax supports both simple

Reviewer: Side-remark: we have a problem with the place `ltac_expr` links to (in the Ltac2 chapter). The grammar chunks extending such tokens in the Ltac2 chapter should probably be converted to use `prodn` with `+=` like in the SSReflect chapter.

Reviewer: Hum, why mark this as resolved?

Reviewer: This is now its own issue.

and compound tactics.  For example: ``split.`` is a simple tactic while ``split; auto.`` combines two
simple tactics.

Tactics specify how to transform the current proof state in order to generate a proof.  They
are syntactically valid only when |Coq| is in proof mode, such as after a :cmd:`Theorem` command
and before any subsequent proof-terminating command such as :cmd:`Qed`.  See :ref:`proofhandling` for more
on proof mode.

By convention, :n:`@command` names begin with uppercase letters, while :n:`@tactic` names begin with lowercase letters.

.. todo:: What about removing the proof part of this grammar from this chapter
          and putting it somewhere where top-level tactics can be described as well?

>           See also #7583.

Reviewer: I love what you wrote in this introduction. I feel like the small look-ahead to the chapters about proof modes, tactics and tactic language is fine, but we could also do without. For now, I would keep this as is.

.. _gallina-assumptions:

Assumptions
-----------

Assumptions extend the environment with axioms, parameters, hypotheses
or variables. An assumption binds an :n:`@ident` to a :n:`@type`. It is accepted
by Coq if and only if this :n:`@type` is a correct type in the environment
preexisting the declaration and if :n:`@ident` was not previously defined in
the same module. This :n:`@type` is considered to be the type (or
specification, or statement) assumed by :n:`@ident` and we say that :n:`@ident`
has type :n:`@type`.

.. _Axiom:

.. cmd:: @assumption_token {? Inline {? ( @num ) } } {| {+ ( @assumpt ) } | @assumpt }
   :name: Axiom; Axioms; Conjecture; Conjectures; Hypothesis; Hypotheses; Parameter; Parameters; Variable; Variables

   .. insertprodn assumption_token of_type

   .. prodn::
      assumption_token ::= {| Axiom | Axioms }
      | {| Conjecture | Conjectures }
      | {| Parameter | Parameters }
      | {| Hypothesis | Hypotheses }
      | {| Variable | Variables }
      assumpt ::= {+ @ident_decl } @of_type
      ident_decl ::= @ident {? @univ_decl }

>       of_type ::= {| : | :> | :>> } @term

Reviewer: Can we use `type` here? Like so:

```suggestion
      of_type ::= {| : | :> | :>> } @type
```

Author: And add a production `type ::= term` in the Terms/Types subsection?

The definition there is not too illuminating:

```
Coq terms are typed. Coq types are recognized by the same syntactic class as term.
We denote by type the semantic subclass of types inside the syntactic class term.
```

Reviewer: You find this definition confusing?

Author: Yes, not very clear at all.  The wording is stilted.  Is this right?
- Every type is a term.
- Not every term is a type??
- Every term has an associated type, but some terms may have the same type (e.g. (S O) and (O) are both of type nat?  If so, how do you get/compute the type of a term?

Reviewer: Everything that you wrote is correct. The typing rules are given in the CIC chapter. The refactoring should make them more accessible.

>    This command binds one or more :n:`@ident`\s to specified :n:`@type`\(s) as their specifications in

Reviewer: ```suggestion
   These commands bind one or more :n:`@ident`\(s) to specified :n:`@type`\(s) as their specifications in
```

>    the global context. The fact asserted by the :n:`@type` is accepted as a

Reviewer: ```suggestion
   the global context. The fact asserted by the :n:`@type` / the existence of an object of this type is accepted as a
```

Author: That sort of makes sense, but is there a precise difference between asserting a fact and asserting the existence of an object?  Or is this saying the same thing in 2 different ways?  Can we give examples.  It's a bit fuzzy with the OR.

   postulate.

   :cmd:`Axiom`, :cmd:`Conjecture`, :cmd:`Parameter` and their plural forms
   are equivalent.  They can take the :attr:`local` attribute (see :ref:`gallina-attributes`),
   which makes the defined :n:`@ident`\s accessible by :cmd:`Import` and its variants
   only through their fully qualified names.

   Similarly, :cmd:`Hypothesis`, :cmd:`Variable` and their plural forms are equivalent.  Outside
   of a section, these are equivalent to :n:`Local Parameter`.  Inside a section, the
   :n:`@ident`\s defined are only accessible within the section.  When the current section
   is closed, the :n:`@ident`\s

>    becomes undefined and every object using this variable will be explicitly

Reviewer: ```suggestion
   become undefined and every object depending on them will be explicitly
```

>    parameterized (i.e., the variable is *discharged*).  See Section :ref:`section-mechanism`.

Reviewer: ```suggestion
   parameterized (i.e., the variables are *discharged*).  See Section :ref:`section-mechanism`.
```

   .. exn:: @ident already exists.
      :name: @ident already exists. (Axiom)
      :undocumented:

   .. warn:: @ident is declared as a local axiom [local-declaration,scope]

>       Warning generated when using :cmd:`Variable` instead of

Reviewer: ```suggestion
      Warning generated when using :cmd:`Variable`, :cmd:`Hypothesis` or their plural forms instead of
```

Author: How about this instead?

Warning generated when using :cmd:`Variable` or its equivalents
      instead of :n:`Local Parameter` or its equivalents.

Reviewer: Sure, that sounds good too.

>       :n:`Local Parameter`.

Reviewer: ```suggestion
      :cmd`Parameter` (or an equivalent variant) with the :attr:`local` attribute.
```

Author: How about `#[local] Parameter` or its equivalent?  Let's be proud of our (funky) syntax!

I tend to think we'll keep `Local` around a while--I expect it's used a lot.

Reviewer: Yeah well, I'm not so proud. Let's keep it `Local Parameter` for now.

.. note::
   We advise using the commands :cmd:`Axiom`, :cmd:`Conjecture` and
   :cmd:`Hypothesis` (and their plural forms) for logical postulates (i.e. when
   the assertion :n:`@type` is of sort :g:`Prop`), and to use the commands
   :cmd:`Parameter` and :cmd:`Variable` (and their plural forms) in other cases

>    (corresponding to the declaration of an abstract mathematical entity).

Reviewer: ```suggestion
   (corresponding to the declaration of an abstract object of the given type).
```

> 

Reviewer: Here is an example that could be added at this point (to show some variants of the syntax):

```coq
Parameter X Y : Set.
Parameter (R : X -> Y -> Prop) (S : Y -> X -> Prop).
Axiom R_S_inv : forall x y, R x y <-> S y x.
```

   decorated-sentence : [ `decoration` … `decoration` ] `sentence`
   sentence           : `assumption`

>                       : `definition`

Author: I wanted your thoughts on where to document proof mode and attributes before I finished revising this part.

There should be some place where we explicitly introduce both commands and tactics and relate them to the rest of the grammar and to the start symbol.  This subsection seems like a good candidate.  (And also note/explain/take into account the fact that the grammar only covers a single command/tactic at a time; the input is a series of start_symbol.  And that you can't have tactics outside proof mode.  And that commands begin with an upper-case letter while tactics do not, etc.)

One alternative is not to have any grammar here at all.  But I think these details should be brought together somewhere.

FWIW, this (from the beginning of "Proof Handling") is overstated.  For example, semantically, you can't open a section within a proof.

"In Coq’s proof editing mode all top-level commands documented in Chapter Vernacular commands remain available"

>                       : `definition`

Reviewer: Attributes are a core part of the grammar, so I feel like it makes sense to document them in this first chapter. Proof modes sort of create extension points. So they do not belong in the early chapters, even though we can still mention the existence of tactics at this point.

             {? Local } Axioms {+ ( {+ @ident } : @type ) }
             {? Local } Conjecture {+ ( {+ @ident } : @type ) }
             {? Local } Conjectures {+ ( {+ @ident } : @type ) }

>       :name: ParametersVar; AxiomVar; AxiomsVar; ConjectureVar; ConjecturesVar

Author: > the main syntax description is quite hard to read...

You mean the syntax itself or the material that follows it?  If you mean the syntax itself, what would you do differently?

I removed the `:name:`s, so this won't be in the index.

>       :name: ParametersVar; AxiomVar; AxiomsVar; ConjectureVar; ConjecturesVar

Reviewer: For instance, I'm annoyed by the presence of the `Inline (num)` optional part. This is apparently specific to the module system, and is related to the undocumented `Inline Level` option (which sets the default inline level when `Inline` is specified without a `(num)`). We have an issue here, which is that commands such as `Parameter` or `Axioms` take a different meaning if they are used at top-level or in module types, and commands such as `Variable` or `Hypothesis` take a different meaning if they are used at top-level or in sections. We should briefly describe this and write that more thorough explanations are to be found in the module / section documentation.

One way of giving less importance to this `Inline` thing would be to move this into an attribute.

>       :name: ParametersVar; AxiomVar; AxiomsVar; ConjectureVar; ConjecturesVar

Reviewer 2: >Parameter or Axioms take a different meaning if they are used at top-level or in module types

Is it really different?

>       :name: ParametersVar; AxiomVar; AxiomsVar; ConjectureVar; ConjecturesVar

Reviewer: I guess that's a matter of viewpoint. They stop being top level assumptions and become something that can be instantiated.

A formal presentation of constants and environments is given in
Section :ref:`typing-rules`.

.. cmd:: {| Definition | Example } @ident_decl @def_body
   :name: Definition; Example

   .. insertprodn def_body def_body

   .. prodn::
      def_body ::= {* @binder } := {? @reduce } @term
      | {* @binder } : @term := {? @reduce } @term

>       | {* @binder } : @term

Reviewer: Can we use `type` here and above? This would make the explanation below easier to understand.

```suggestion
      | {* @binder } : @type
```

   These commands bind :n:`@term` to the name :n:`@ident` in the environment,
   provided that :n:`@term` is well-typed.  They can take the :attr:`local` attribute (see :ref:`gallina-attributes`),
   which makes the defined :n:`@ident`\s accessible by :cmd:`Import` and its variants

>    only through their fully qualified names.

Reviewer: ```suggestion
   only through their fully qualified names.
   If :n:`@reduce` is present then :n:`@ident` is bound to the result of the specified
   computation on :n:`@term`.
```

A formal presentation of constants and environments is given in
Section :ref:`typing-rules`.

.. cmd:: {| Definition | Example } @ident_decl @def_body
   :name: Definition; Example

   .. insertprodn def_body def_body

   .. prodn::
      def_body ::= {* @binder } := {? @reduce } @term
      | {* @binder } : @type := {? @reduce } @term

>       | {* @binder } : @type

Reviewer: This could be simplified into:

```

def_body ::= {* @binder } {? : @type } := {? @reduce } @term
| {* @binder } : @type
```

   These commands bind :n:`@term` to the name :n:`@ident` in the environment,
   provided that :n:`@term` is well-typed.  They can take the :attr:`local` attribute (see :ref:`gallina-attributes`),
   which makes the defined :n:`@ident` accessible by :cmd:`Import` and its variants
   only through their fully qualified names.
   If :n:`@reduce` is present then :n:`@ident` is bound to the result of the specified
   computation on :n:`@term`.

>    If :n:`@term` is omitted, Coq enters the proof editing mode.  This is useful WHY EXACTLY?

Reviewer: This can be used to define a term incrementally, in particular by relying on the `refine` tactic.

   In this case, the proof should be terminated with :cmd:`Defined` in order to define a constant
   for which the computational behavior is relevant.  See :ref:`proof-editing-mode`.

   .. seealso:: :cmd:`Opaque`, :cmd:`Transparent`, :tacn:`unfold`.

   .. exn:: @ident already exists.
      :name: @ident already exists. (Definition)
      :undocumented:

>       The form :n:`Definition @ident : @type := @term` checks that the type of :n:`@term`

Reviewer: This is improperly indented as if it was documenting the error message.

A formal presentation of constants and environments is given in
Section :ref:`typing-rules`.

.. cmd:: @def_token @ident_decl @def_body
   :name: Definition; Example; SubClass

   .. insertprodn def_token pattern_occ

   .. prodn::
      def_token ::= Definition
      | Example
      | SubClass
      def_body ::= {* @binder } := {? @reduce } @term
      | {* @binder } : @term := {? @reduce } @term
      | {* @binder } : @term
      reduce ::= Eval @red_expr in

>       red_expr ::= red

Author: I'll break out the grammar for red_expr separately.

I don't know precisely where the dividing line between the kernel and the "elaboration engine" lies.

> I'm also a bit worried how much the objective of describing the complete syntax of commands at once conflicts with the refactoring objective of first describing the kernel, then the elaboration engine

Reality is messier than our neat conceptions.  The refactoring is not yet started.  (Indeed, we've not even put out the CEP plan for doing so--which I hope we can do within, say, the next month.)

A useful way to think about **any** decision is how difficult/impactful it would be to reverse.  If it's difficult to do so then you should do more analysis up front.  If it's easy, then much less analysis.  Seems to me it would be quite easy to move a `red_expr` subsection between chapters.

>       red_expr ::= red

Author: If you want `red_expr` elsewhere I will put it there now.

>       red_expr ::= red

Reviewer: For now, let's have a new subsection "Computation" say between "Assertions" and "Attributes" where you can move the grammar from `reduce` to `pattern_occ`.

We gradually explain simple inductive types, simple annotated inductive
types, simple parametric inductive types, mutually inductive types. We
also explain co-inductive types.

Simple inductive types
~~~~~~~~~~~~~~~~~~~~~~

> .. cmd:: Inductive @inductive_definition {* with @inductive_definition }

Author: I get your point but I'm not sure where to reinsert "We gradually..." and "Simple inductive types".  Seems like this may need some new text.

> .. cmd:: Inductive @inductive_definition {* with @inductive_definition }

Reviewer: > Seems like this may need some new text.

Agreed. Do put something that makes sense even if not perfect. I can try to help making it better, but overall the changes that this PR brings are worth degrading the text a little.

   .. insertprodn inductive_definition record_binder_body

   .. prodn::
      inductive_definition ::= {? > } @ident_decl {* @binder } {? : @term } {? := {? @constructor_list_or_record_decl } } {? @decl_notations }

>       constructor_list_or_record_decl ::= %| {+| @constructor }

Author: > Also, the leading slash should be optional.

Huh?? "slash" is "/", right?

>       constructor_list_or_record_decl ::= %| {+| @constructor }

Reviewer: Yeah. I meant pipe.

>       constructor_list_or_record_decl ::= %| {+| @constructor }

Author: The pipe is unconditional in the code.  fullGrammar has:

```
constructor_list_or_record_decl: [
| "|" LIST1 constructor SEP "|"
   : 
```
Perhaps you would submit an issue for this?

>       constructor_list_or_record_decl ::= %| {+| @constructor }

Reviewer: The full list of rules is:

```
constructor_list_or_record_decl: [
| "|" LIST1 constructor SEP "|"
| identref constructor_type "|" LIST0 constructor SEP "|"
| identref constructor_type
| identref "{" record_fields "}"
| "{" record_fields "}"
|
]
```

The second and third rules show that the leading pipe is actually optional.

>       constructor_list_or_record_decl ::= %| {+| @constructor }

Reviewer: I'm currently fixing the issue with the trailing pipe. The reason why this simple rule is split in three and sometimes `constructor` is inlined and sometimes it is not is to disambiguate the case where there is a leading `identref` as it could be followed by a record or a `constructor_type`.

      | @ident @constructor_type %| {*| @constructor }

>       | @ident @constructor_type

Author: I don't see how to reduce it to a single production.  The grammar allows "c", "| c" and "c |" but not "| c |" nor "c | c |" where c = ident constructor_type.  Note that the first production uses + while the second uses *.  Maybe the syntax should be changed?

I made some partial changes, though.

>       | @ident @constructor_type

Reviewer: I don't think that `c |` is supported. Where do you see this in the grammar?

Assuming that `constructor ::= @ident @constructor_type`, then what about:

```
constructor_list_or_record_decl ::= {? %| } {+| @constructor }
```

Which case am I missing?

>       | @ident @constructor_type

Author: Second production with zero repetitions of the *.

>       | @ident @constructor_type

Author: Second production with zero repetitions of the *.

>       | @ident @constructor_type

Reviewer: Oh right! I opened #11654. I think we'll remove the syntax and thus should be able to simplify the rules.

>       | @ident @constructor_type

Reviewer: When #11660 is merged, the three rules can indeed be merged into `{? %| } {+| @constructor }`.

      | @ident %{ {+; @record_field } %}
      | %{ {+; @record_field } %}
      constructor_type ::= {* @binder } {? @of_type }
      record_field ::= {* #[ {*, @attr } ] } @record_binder {? %| @num } {? @decl_notations }
      record_binder ::= @name
      | @name @record_binder_body

>       record_binder_body ::= {* @binder } @of_type

Author: `record_body` doesn't seem right, either.  It's a subpart of the definition of a single `record_field`.  Maybe it could be `field_body`?  That still echos `def_body`.

There's also already a definition for `record_body` in the gallina extensions chapter that this conflicts with.  I'd rather not pull that apart now since there may be rippling changes from removing that.

>       record_binder_body ::= {* @binder } @of_type

Author: I made it `field_body` for now.  At least that way CI will generate HTML output.

>       record_binder_body ::= {* @binder } @of_type

Reviewer: Great!

>       record_binder_body ::= {* @binder } @of_type

Reviewer 3: > why records do not support the `Eval` construct

No particular reasons. Nobody was feeling it was missing probably, plus, maybe, the fact that the use of a capital letter for `Eval` does not look very nice visually.

If you want to implement it, it'll have to be passed from `g_vernac.mlg` to `Record.definition_structure` via `vernacentries.ml` with a call to `red_constant_body` after interpretation, following the model done in `ComDefinition.interp_definition`.

      | {* @binder } @of_type := @term
      | {* @binder } := @term

>    This command defines a simple inductive type and its constructors.

Author: Well, what used to be there for `cmd::` was incorrect, e.g. here's no @sort in the current grammar.

Probably what's needed here is an introductory taxonomy of inductive definition types and why/when you use each.  I understand mutually defined inductive types well enough to describe, but the others not so much.  Not entirely clear on which types/attributes are mutually exclusive. (e.g. can you have a mutually defined annotated inductive type?)  Seems like it would be a bit of a deep dive, so I would prefer to defer most of this to another PR.

Isn't the definition of "simple inductive type" more or less "any inductive type that doesn't use features A, B, C, ..."?  So you need to understand all of the alternatives up front.

Your thoughts?

>    This command defines a simple inductive type and its constructors.

Reviewer: > there's no `@sort` in the current grammar.

There's `{? : @type }`. In the simple inductive case, `@type` is restricted to being a `@sort`.

>Not entirely clear on which types/attributes are mutually exclusive.

There are no mutual exclusive features. These really are more examples rather than a taxonomy. I.e. the original authors just wanted to introduce inductive types progressively rather than in full generality. We are taking the opposite path.

>Isn't the definition of "simple inductive type" more or less "any inductive type that doesn't use features A, B, C, ..."?

Right. So this is a good first example, but it shouldn't be defined negatively. It should just be presented as the simplest example of an inductive type.

>    This command defines a simple inductive type and its constructors.

Author: I was mistaken:`@sort` is defined in the grammar, though not visible in these productions.

The old cmd is `Inductive @ident : {? @sort } := {? | } @ident : @type {* | @ident : @type }`.  However, `: {? @sort }` should be `{? : @sort }`, then it should be simplified to:
`Inductive @ident {? : @sort } := {? | } {+| @ident : @type }`.  Also, the 

However, the relationship between this and the new syntax is not as obvious as I'd like:
- Using `@sort` in the old and `@type` in the new requires a sentence of explanation
- The old cmd drops `{? > } ... {* @binder } {? %| {* @binder } } ...  {? @decl_notations }` that are present in the new without explanation.
- The old syntax is very selective about what it includes from `constructors_or_record`
- `:=` is made unconditional while things in the old form remain conditional, such as `@sort`, which the text assumes is present.  Which should it be?

I suppose the text could say something like "Simple inductive types use a :n:`@type` that is a :n:`sort`.  <syntax notation> represents this in simplified form.", but yuck.  It would be a whole lot simpler just to begin with the `nat` example below and avoid showing a subset of the syntax in notation form at all.  (And probably avoid showing subset syntax anywhere in the doc.)

(IMHO this section needs a good bit more attention, but not in this PR.)

>    This command defines a simple inductive type and its constructors.

Reviewer: I agree with your analysis.

.. cmd:: Inductive @ident : {? @sort } := {? | } @ident : @type {* | @ident : @type }

   This command defines a simple inductive type and its constructors.
   The first :n:`@ident` is the name of the inductively defined type
   and :n:`@sort` is the universe where it lives. The next :n:`@ident`\s
   are the names of its constructors and :n:`@type` their respective types.
   Depending on the universe where the inductive type :n:`@ident` lives
   (e.g. its type :n:`@sort`), Coq provides a number of destructors.

>    Destructors are named :n:`@ident`\ ``_sind``, :n:`@ident`\ ``_ind``,

Author: Can't do that, `@ident_sind` reports that `ident_sind` is an undefined token.

>    Destructors are named :n:`@ident`\ ``_sind``, :n:`@ident`\ ``_ind``,

Reviewer: Oops, sorry for the bad advice!

>    :n:`@ident`\ ``_rec`` or :n:`@ident`\ ``_rect`` which respectively

Author: Nor this.

   correspond to elimination principles on :g:`SProp`, :g:`Prop`, :g:`Set` and :g:`Type`.
   The type of the destructors expresses structural induction/recursion
   principles over objects of type :n:`@ident`.

>    The constant :n:`@ident`\ ``_ind`` is always provided,

Author: nor this

>    whereas :n:`@ident`\ ``_rec`` and :n:`@ident`\ ``_rect`` can be

Author: nope

   .. exn:: The conclusion of @type is not valid; it must be built from @ident.

      The conclusion of the type of the constructors must be the inductive type
      :n:`@ident` being defined (or :n:`@ident` applied to arguments in
      the case of annotated inductive types — cf. next section).

The following subsections show examples of simple inductive types, simple annotated
inductive types, simple parametric inductive types and mutually inductive
types.

.. _simple-inductive-types:

Simple inductive types
~~~~~~~~~~~~~~~~~~~~~~

> A simple inductive type belongs to a universe that is a simple :n:`sort`.

Author: What does it mean to be a "simple sort"?  Is that different from @sort?  If so, that should be defined--I can do that in the other PR.  And :n:\`sort\` should be :n:\`@sort`

Reviewer: It's just a sort. I think it is written a simple sort as opposed to a complex type.

      You have to explicitly give their fully qualified name to refer to them.

   .. cmdv:: {? Local } Example @ident {? @binders } {? : @type } := @term
      :name: ExampleVar

      This is equivalent to :cmd:`Definition`.

>    .. cmdv:: Let @ident @def_body

Author: It is indeed.  Would you mind opening the PR?  I don't know enough about this to write a good PR.

In fullGrammar:

```
| "Let" identref def_body
```

>    .. cmdv:: Let @ident @def_body

Reviewer: See #11534

   are the names of its constructors and :token:`type` their respective types.
   Depending on the universe where the inductive type :token:`ident` lives
   (e.g. its type :token:`sort`), Coq provides a number of destructors.

>    Destructors are named :token:`ident`\ ``_sind``, :token:`ident`\ ``_ind``,

Author: There are quite a few of these in the rst files.  After fixing 139 in this file, that still leaves some 600 uses in the other files.  Probably do those in another PR.

   be defined using :cmd:`Variant`). No induction scheme is generated for
   this variant, unless the :flag:`Nonrecursive Elimination Schemes` flag is on.

>    This command supports the same attributes as the :cmd:`Inductive` command.

Reviewer: We can go back to the original wording, but then the list of supported attributes should be kept.

.. _Fixpoint:

> .. cmd:: Fixpoint @fix_definition {* with @fix_definition }

Reviewer: This includes the syntax for mutually recursive definitions but no corresponding explanation is given here, only in the `cmdv` below. What can we do about this? We should either, include the explanation on mutually recursive definitions sooner, or change this to `Fixpoint @fix_definition` and leave the full syntax for the variant below.

Author: Should this be as shown, or maybe `{? : type } {? := term }`?
![image](https://user-images.githubusercontent.com/1253341/74711222-50e21100-51d8-11ea-8574-4a40dc2bd4ce.png)

Author: I've not tried to clean up all the cmdv's--seemed like you might have to clean up a lot of text.  And understand it in some depth.

This one is partial:

![image](https://user-images.githubusercontent.com/1253341/74711455-dc5ba200-51d8-11ea-9cff-922b36bd24ae.png)

- Missing part of fixannot
- := term is optional

Author: Also this seems weird: "This command allows defining functions by pattern matching over inductive objects using a fixed point construction." for which it seems a `match` is required, but the syntax doesn't cover this (presumably it's a restriction on `@term`).  Not sure what to say here.

Reviewer: - Yes, I was suggesting the use of `{? : type }`.

- It would be great if we could take a decision on whether to remove or to clean this particular `cmdv`.

- There is no syntactic restriction and indeed, you can use `Fixpoint` without a `match` like so: `Fixpoint foo (n : nat) := 0.` (this is not actually useful because there is no recursive call). The verification that the recursive calls are done on structurally smaller arguments is enough to generally imply that `term` will contain a `match` construct.

Author: No syntactic restriction, nor (based on your example) is there any semantic restriction requiring `match`, right?

Reviewer: Right. It's just that you need a `match` to write any meaningful recursive function.

   .. insertprodn fix_definition decl_notation

   .. prodn::

>       fix_definition ::= @ident_decl {* @binder } {? @fixannot } {? : @term } {? := @term } {? @decl_notations }

Reviewer: Would be better to change this to:

```suggestion
      fix_definition ::= @ident_decl {* @binder } {? @fixannot } {? : @type } {? := @term } {? @decl_notations }
```

so that the description below makes more sense.

      decl_notations ::= where @decl_notation {* and @decl_notation }
      decl_notation ::= @string := @term1_extended {? : @ident }

   This command allows defining functions by pattern matching over inductive
   objects using a fixed point construction. The meaning of this declaration is
   to define :n:`@ident` a recursive function with arguments specified by

>    the :n:`@binders` such that :n:`@ident` applied to arguments

Reviewer: Everywhere there is a reference to `binders` it should become ```:n:`@binder`\s``` since the latter does not exist anymore.

Author: It would helpful if Sphinx generated an error message for this type of missing reference

Reviewer: It does though, but you have disabled it by adding `binders` to the `nitpick_ignore` list in `conf.py`.

.. _Fixpoint:

.. cmd:: Fixpoint @fix_definition {* with @fix_definition }

   .. insertprodn fix_definition fix_definition

   .. prodn::
      fix_definition ::= @ident_decl {* @binder } {? @fixannot } {? : @type } {? := @term } {? @decl_notations }

   This command allows defining functions by pattern matching over inductive
   objects using a fixed point construction. The meaning of this declaration is
   to define :n:`@ident` as a recursive function with arguments specified by
   the :n:`@binder`\s such that :n:`@ident` applied to arguments
   corresponding to these :n:`@binder`\s has type :n:`@type`, and is
   equivalent to the expression :n:`@term`. The type of :n:`@ident` is
   consequently :n:`forall {* @binder }, @type` and its value is equivalent
   to :n:`fun {* @binder } => @term`.

   To be accepted, a :cmd:`Fixpoint` definition has to satisfy some syntactical
   constraints on a special argument called the decreasing argument. They
   are needed to ensure that the :cmd:`Fixpoint` definition always terminates.

>    The point of the :n:`{struct @ident}` annotation is to let the user tell the

Reviewer: It would be great to note the link with the `@fixannot` nonterminal in the grammar above.

   system which argument decreases along the recursive calls.

   The :n:`{struct @ident}` annotation may be left implicit, in which case the
   system successively tries arguments from left to right until it finds one
   that satisfies the decreasing condition.

   The :n:`with` clause allows simultaneously defining several mutual fixpoints.
   It is especially useful when defining functions over mutually defined
   inductive types.  Example: :ref:`Mutual Fixpoints<example_mutual_fixpoints>`.

   If :n:`@term` is omitted, :n:`@type` is required and Coq enters the proof editing mode.

>    This can be used to define a term incrementally, in particular by relying on the refine tactic.

Reviewer: ```suggestion
   This can be used to define a term incrementally, in particular by relying on the :tacn:`refine` tactic.
```

.. _Fixpoint:

.. cmd:: Fixpoint {+with @fix_definition }

   .. insertprodn fix_definition decl_notation

   .. prodn::

>       fix_definition ::= @ident_decl {* @binder } {? @fixannot } {? : @term } {? := @term } {? @decl_notations }

Reviewer: Interesting! The following is syntactically correct:

```
Fixpoint foo.
```

However, it raises an error nonetheless: "A fixpoint needs at least one parameter".

This contrasts with the grammar of `Definition` where the choice that was made was to eliminate a definition without body nor type at parsing time.

Definitions of recursive objects in co-inductive types
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. cmd:: CoFixpoint @cofix_definition {* with @cofix_definition }

   .. insertprodn cofix_definition cofix_definition

   .. prodn::

>       cofix_definition ::= @ident_decl {* @binder } {? : @term } {? := @term } {? @decl_notations }

Reviewer: Same remarks on `CoFixpoint`.

Author: I removed the cmdv.  I reworded but didn't move the statement after it.

      As in the :cmd:`Fixpoint` command, it is possible to introduce a block of
      mutually dependent methods.

.. _Computation:

> Computation

Reviewer: I suggest the plural form for uniformity of section titles:
```suggestion
Computations
```

-----------

.. insertprodn reduce pattern_occ

.. prodn::
   reduce ::= Eval @red_expr in
   red_expr ::= red
   | hnf
   | simpl {? @delta_flag } {? @ref_or_pattern_occ }
   | cbv {? @strategy_flag }
   | cbn {? @strategy_flag }
   | lazy {? @strategy_flag }
   | compute {? @delta_flag }
   | vm_compute {? @ref_or_pattern_occ }
   | native_compute {? @ref_or_pattern_occ }
   | unfold {+, @unfold_occ }
   | fold {+ @term1_extended }
   | pattern {+, @pattern_occ }
   | @ident
   delta_flag ::= {? - } [ {+ @smart_global } ]
   smart_global ::= @qualid
   | @by_notation
   by_notation ::= @string {? % @ident }
   strategy_flag ::= {+ @red_flags }
   | @delta_flag
   red_flags ::= beta
   | iota
   | match
   | fix
   | cofix
   | zeta
   | delta {? @delta_flag }
   ref_or_pattern_occ ::= @smart_global {? at @occs_nums }
   | @term1_extended {? at @occs_nums }
   occs_nums ::= {+ @num_or_var }
   | - @num_or_var {* @int_or_var }
   num_or_var ::= @num
   | @ident
   int_or_var ::= @int
   | @ident
   unfold_occ ::= @smart_global {? at @occs_nums }
   pattern_occ ::= @term1_extended {? at @occs_nums }

> Details TBA.

Reviewer: Instead, you can just put a link to the "Conversion rules" section of the CIC chapter and a `.. todo::`.

      Eval compute in (hd (from 0)).
      Eval compute in (tl (from 0)).

   As in the :cmd:`Fixpoint` command, the :n:`with` clause allows simultaneously
   defining several mutual cofixpoints.

   If :n:`@term` is omitted, :n:`@type` is required and Coq enters the proof editing mode.

>    This can be used to define a term incrementally, in particular by relying on the refine tactic.

Reviewer: ```suggestion
   This can be used to define a term incrementally, in particular by relying on the :tacn:`refine` tactic.
```

proof mode is described in Chapter :ref:`proofhandling` and the tactics in
Chapter :ref:`Tactics`. The basic assertion command is:

> .. cmd:: @thm_token @ident_decl {* @binder } : @term

Reviewer: Either `term` should be replaced by `type` or the explanation below should be amended.

   :name: Theorem; Lemma; Fact; Remark; Corollary; Proposition

>    .. ; Property

Reviewer: Why is `Property` left out of the index?

   .. insertprodn thm_token thm_token

   .. prodn::
      thm_token ::= Theorem
      | Lemma
      | Fact
      | Remark
      | Corollary
      | Proposition
      | Property

   After the statement is asserted, Coq needs a proof. Once a proof of
   :n:`@type` under the assumptions represented by :n:`@binders` is given and
   validated, the proof is generalized into a proof of :n:`forall @binders, @type` and
   the theorem is bound to the name :n:`@ident` in the environment.

>    .. exn:: The term @term has type @type which should be Set, Prop or Type.

Reviewer: OTOH this error message is consistent with the current state of the assertion syntax.

proof mode is described in Chapter :ref:`proofhandling` and the tactics in
Chapter :ref:`Tactics`. The basic assertion command is:

> .. cmd:: @thm_token @ident_decl {* @binder } : @type {* with @ident {* @binder } : @type}

Author: It does.  Editing error.

      This feature, called nested proofs, is disabled by default.
      To activate it, turn the :flag:`Nested Proofs Allowed` flag on.

> .. cmd:: thm_token @ident {? @binders } : @type {* with @ident {? @binders } : @type}

Reviewer: This used to be a `cmdv`. Now that it is a `cmd`, it should be moved below the `Definition` and `Let` variants, just above the `Fixpoint` and `CoFixpoint` variants.

Author: Mixed feelings about this.  You wanted to split this in 2, only showing the `with` clause in the second one.  But it's really good to keep all the syntax together.  And I'd like to get rid of all the cmdvs, if possible.  Take a look at my latest change--do you think that's workable

proof mode is described in Chapter :ref:`proofhandling` and the tactics in
Chapter :ref:`Tactics`. The basic assertion command is:

> .. cmd:: @thm_token @ident_decl {* @binder } : @term {* with @ident_decl {* @binder } : @term }

Reviewer: This is a case where the use of the `with` clause changes the semantics so significantly that we should really document the two variants separately (as separate commands, with separate indexes).

I've tried in the past to deprecate `Theorem ... with` without success (see #8265). More generally, I've tried to make the syntax of `Definition` and `Theorem` closer (see #6567 and the plan to support the `Theorem ... := ...` syntax), but did not make much progress so far...

Author: This applies not just `Theorem ... with` but to `Lemma, ...` as well?

Reviewer: Yes

``#[foo]#[bar]`` and ``#[bar]#[foo]`` are equivalent.

The legacy attributes (:n:`@legacy_attrs`) provide an older, alternate syntax
for certain attributes.

Some attributes are specific to a command, and so are described with
that command. Currently, the following attributes are recognized by a
variety of commands:

.. attr:: universes(monomorphic)
   :name: universes(monomorphic)

   See :ref:`polymorphicuniverses`.

.. attr:: Monomorphic

   Legacy attribute equivalent to :attr:`universes(monomorphic)`.

.. attr:: universes(polymorphic)
   :name: universes(polymorphic)

   See :ref:`polymorphicuniverses`.

.. attr:: Polymorphic

   Legacy attribute equivalent to :attr:`universes(polymorphic)`.

.. attr:: program

   See :ref:`programs`.

.. attr:: Program

   Legacy attribute equivalent to :attr:`program`.

.. attr:: global

   See :ref:`controlling-locality-of-commands`.

.. attr:: Global

   Legacy attribute equivalent to :attr:`global`.

.. attr:: local

   See :ref:`controlling-locality-of-commands`.

.. attr:: Local

   Legacy attribute equivalent to :attr:`local`.

.. attr:: Cumulative

>    Legacy attribute THAT DOES WHAT?  Allowed only in a polymorphic context.

Reviewer: See https://coq.gitlab.io/-/coq/-/jobs/442127533/artifacts/_install_ci/share/doc/coq/sphinx/html/addendum/universe-polymorphism.html#coq:cmd.cumulative. I am quite surprised that there is no `universes(cumulative)` syntax for this.

Note that `Cumulative`, `NonCumulative` and `Private` do not satisfy the above description of being "recognized by a variety of commands" (unless we consider that `Inductive` / `Record` / `Structure` / `Variant` represent a variety of commands, in which case this list is missing out on `universes(template)` and `universes(notemplate)`).

Author: I don't understand universe polymorphism or cumulativity well enough to write up for the manual.  I don't see where they are defined with any clarity in the manual.  The `Cumulative` command is only described as "Declares the inductive as cumulative".  Indeed, I only have a vague understanding of universes.

> Note that `Cumulative`, `NonCumulative` and `Private` do not satisfy the above description of being "recognized by a variety of commands"

True.  My inclination has been to deem any flag that appears before the main command name as an attribute.  That has its plusses and minuses for readability (it's in the way if we show {? Cumulative | NonCumulative } in the main syntax, but it's somewhat buried if the reader has to look at the following text to see that the command does indeed take certain attributes).  I could go either way--though we already made a decision on this and it would be some wasted motion to change it.

Reviewer 4: > I don't see where they are defined with any clarity in the manual

The section @Zimmi48 linked to seems pretty clear to me.  Note that there is a large block of text with a couple of examples describing cumulativity; if you keep reading up to the heading "Cumulativity Weak Constraints", that is all still about cumulativity.  In any case, the description of NonCumulative is quite easy: it says that for two instances of the same inductive type (family) to be convertible, all the universes must be equal.  Cumulative relaxes this constraint according to the variance of the universes.

Author: I can use what you said for NonCumulative.

For Cumulative, variance is "invariant, covariant or irrelevant".  How are each of these treated for conversion (in the simplest language possible)?  I would rather not guess.

Reviewer 4: I'm surprised there is no contravariant.
Irrelevant means that conversation of inductive type (families) imposes no constraint on this universe.  See also the list example in the docs.
I believe invariant imposes an equality constraint.  I expect that `Inductive foo@{i} := bar (_ : Type{i} -> Type@{i})` would make `i` invariant, and then equating `foo@{i}` and `foo@{j}` judgmentally would equate `i` and `j`.  You should perhaps check this, and maybe the docs should be updated with this example (or a less contrived one).
Covariant will clearly impose a `<=` constraint, but directionally is more subtle.  Let me check the docs.

Reviewer 4: Btw, "Cic" seems like the wrong capitalization in https://coq.gitlab.io/-/coq/-/jobs/442127533/artifacts/_install_ci/share/doc/coq/sphinx/html/language/cic.html#subtyping-rules , maybe we should open an issue about it?

Reviewer 4: As far as I can tell, the section on conversation rules has not been updated to account for cumulative universes.  Here is my guess for covariance, which should be tested:
In addition to getting the conversion problem that am inductive family must be equal at two different universe instantiations, you can also get the conversation problem that a term in the inductive family at one universe instance must also live in the inductive at another instance.  Consider, for example, the conversation problem arising from `fun x : T@{i} => (x : T@{j})` for some inductive `T`.  If the universe in `T` is irrelevant, then this imposes no constraint, as in the funext example involving `eq`.  If the universe is invariant, then this imposes the constraint `i = j`.  If, however, the universe is covariant, that this imposes the constraint `i <= j`.  The prototypical example of an inductive with a covariant universe is `Record T@{i} := { ty : Type@{i} }.` A sightly more interesting example is `Record Dyn@{i} := { typ : Type@{i} ; val : typ }`.  Note that the universe argument to `Type` is covariant.

This is all me making educated guesses; you (or someone) should check these examples before taking my word for it, and/or get one of the devs responsible for universes ( @mattam82 or @skyskimmer or @amintimany , I believe) to confirm.

Also, if I'm correct, these examples should be added to the chapter on cumulatively

Author: > Btw, "Cic" seems like the wrong capitalization in ...., maybe we should open an issue about it?

Yeah, since CIC is an acronym it would normally be in all caps.

.. attr:: NonCumulative

   Legacy attribute THAT DOES WHAT?  Allowed only in a polymorphic context.

.. attr:: Private

>    Legacy attribute that DOES WHAT?

Reviewer: Private inductive types are an undocumented (yet critical to some users) feature. See #11223.
